File: python_parallelpi/parallelpi.py
from multiprocessing import Pool
import math
def chunk(args):
    start = int(args[0])
    end = int(args[1])
    accum = 0
    sign = 1 if start % 2 == 0 else -1
    for i in range(start, end):
        denominator = i * 2 + 1
        accum = accum + (1 / denominator) * sign
        sign = sign * -1
    return accum
def multi():
    accum = 0
    sign = 1
    start = 0
    end = 1000000000
    arguments = []
    numprocessor = 8
    for i in range(0, numprocessor):
        arguments.append((i * end / numprocessor, (i + 1) * end / numprocessor))
    with Pool(processes=numprocessor) as p:
        results = p.map(chunk, arguments)

    accum = sum(results)
    pi = accum * 4
    print('pi is ' + str(pi))

def single():
    start = 0
    end = 10000
    accum = 0
    sign = 1 if start % 2 == 0 else -1
    for i in range(start, end):
        denominator = i * 2 + 1
        accum = accum + (1 / denominator) * sign
        sign = sign * -1

    pi = accum * 4
    print('pi is ' + str(pi))


if __name__ == "__main__":
    multi()
    # single()

File: go_parallelpi/parallelpi.go
package main

import (
	"fmt"

	"github.com/exascience/pargo/parallel"
)

func single() {
	var start int = 0
	var end int = 100
	var accum float64 = 0

	var sign float64 = -1
	if start%2 == 0 {
		sign = 1
	}
	var denominator float64
	for i := start; i < end; i++ {
		denominator = float64(i*2 + 1.0)
		accum = accum + float64(1.0/denominator*sign)
		sign = sign * -1
	}

	var pi float64 = accum * 4
	fmt.Printf("Prime number is %f", pi)

}
func multi() float64 {
	var accum float64 = 0
	var start int = 0
	var end int = 1000000000
	var numprocessor int = 8

	accum = parallel.RangeReduceFloat64(
		start, end, numprocessor,
		func(start, high int) float64 {
			var sign float64 = -1
			var accum float64 = 0
			if start%2 == 0 {
				sign = 1
			}
			var denominator float64
			for i := start; i < end; i++ {
				denominator = float64(i*2 + 1.0)
				accum = accum + float64(1.0/denominator*sign)
				sign = sign * -1
			}
			return accum
		},
		func(left, right float64) float64 {
			return left + right
		},
	)

	var pi float64 = accum * 4
	fmt.Printf("Prime number is %f", pi)
	return accum
}

func main() {
	// single()
	multi()
}
